### DISCUSSION TITLE
Multiple SPI and I2C and UART ports on the STM32

### DISCUSSION THREAD

**Discussion by Semsem-220109** [stm32duino/Arduino_Core_STM32#1981] [Libraries] [answered]
Hi there,
I have a project where I need to make a rc remote transmitter using the STM32F103C8T6 and the NRF24L01. At first, I connected the NRF24L01's SPI pins to the pins PA5 PA6 PA7 wich appears to be the STM32's main SPI pins, but then I realized that those pins are also ADC inputs that I need to use, so I thought about switching the NRF module to the SPI2 (PB13 PB14 PB15), but how to do that in my code in the the Arduino IDE since the NRF24L01 library automatically uses the main SPI port and I believe does not have any command to choose the SPI port ? 

Also, I need to have an I2C OLED display and a MPU6050 wich also uses an I2C communication, so I wanted to have the display connected to the I2C1 (PB6 PB7) and the IMU to I2C2 (PB10 PB11). How to do that ?

Here is the pinout that I'm using to choose the pins to use :
![image](https://user-images.githubusercontent.com/121195717/227335094-5c875455-584c-4b37-97f5-2399d0685737.png)

**Comment by fpistm** [comment]
Fo SPI, simply change default pins: https://github.com/stm32duino/Arduino_Core_STM32/wiki/API#change-default-spi-instance-pins

For I2C also read the wiki.

**Reply by Semsem-220109** [reply]
Then I can use the NRF24L01 library for example normally ?
I see that in the setup, there are some data sent from the SPI_3, do I have to remove them ? What should I keep for just changing the SPI default pins ?
Can you please precisely tell me how to change the SPI port to SPI2 to get the NRF24L01 library working on it ?

**Reply by fpistm** [reply]
> Then I can use the NRF24L01 library for example normally ? 
Default `SPI `instance uses thoses pins:
https://github.com/stm32duino/Arduino_Core_STM32/blob/5a3f412d385c386eb421ede38c4223dbab5d83df/variants/STM32F1xx/F103C8T_F103CB(T-U)/variant_PILL_F103Cx.h#L102-L114
So if you as it seems the NRF library uses default `SPI `instance and you want use other pins, simply change the default one using the API provided befaore the call of `SPI.begin()` or the NRF one: 
```
SPI.setMISO(PB14);
setMOSI(PB15);
SPI.setSCLK(PB13);
```

> I see that in the setup, there are some data sent from the SPI_3, do I have to remove them ?

I don't know what you talk about as you do not share your sketch so can't answer.

> What should I keep for just changing the SPI default pins ? Can you please precisely tell me how to change the SPI port to SPI2 to get the NRF24L01 library working on it ?

See above.

You have all anwsers in the wiki with examples.

**Reply by Semsem-220109** [reply]
Thank you very much, I'll try it when I get home 2 days later :)
And one last question, when to use PB15 and when to use PB_15 ?

**Reply by fpistm** [reply]
Use PB15 which is pin number used for all Arduino API

**Reply by Semsem-220109** [reply]
And how can I use 2 different SPI simultaneously? For example using the default SPI for a NRF24L01 and another SPI for a RFID module ?

**Reply by fpistm** [reply]
Simply declare a second SPI instance on an other SPIx peripheral.

`SPIClass SPI_x(MOSI pin, MISO pin, SCLK pin);`

**Reply by Semsem-220109** [reply]
Ok thanks :) and then I call SPIx.begin(); or SPI_x.begin(); ?

**Reply by fpistm** [reply]
SPI_x is an example. Name it as you want.

**Reply by Semsem-220109** [reply]
Ok thank you

**Reply by Semsem-220109** [reply]
For example I've used the first SPI port for the NRF24L01 with RF24.begin() for example. For example I want to use the SPI2 for another component with component.begin(); for example. How do I do that ? And how can I use multiple USART ? Do I declare them with the hardwareserial library ?

**Reply by fpistm** [reply]
https://github.com/stm32duino/Arduino_Core_STM32/wiki/API#hardwareserial

**Reply by Semsem-220109** [reply]
What about the SPI ?

**Reply by fpistm** [reply]
The same. Instantiate another one as stated before.

**Reply by Semsem-220109** [reply]
Yes but if I put component.begin(); how do I choose the SPI bus ? I think that by default it will select SPI1 which I assigned to the NRF24L01, so I want to use SPI2

**Reply by fpistm** [reply, marked as answer]
Well this is linked to the component library which do not allow this. So you have to use the same SPI with different CS or modify the library to simply use the second SPI.

**Reply by Semsem-220109** [reply]
There's no link

**Reply by fpistm** [reply]
no link on what? What I said is the third party library you used for your component seems not allowing to change default SPI instance. No link with the core.

**Reply by Semsem-220109** [reply]
Sorry I read the message above wrong, I thought you said this is the link for the component library... and there was no link
